File: typescript/patterns-use-cases/microservices-sagas/src/services/flights.ts
import * as restate from "@restatedev/restate-sdk";

export const flights = restate.service({
    name: "flights",
    handlers: {
      reserve: async (ctx: restate.Context, tripId: string) => {
        const flightBooking = ctx.rand.uuidv4();
        console.info(`Flight ${flightBooking} reserved for trip ${tripId}`);
        return flightBooking;
      },

      confirm: async (ctx: restate.Context, req: {tripId: string, flightBooking: string}) => {
        console.info(`Flight ${req.flightBooking} confirmed for trip ${req.tripId}`);
      },

      cancel: async (ctx: restate.Context, req: {tripId: string, flightBooking: string}) => {
        console.info(`Flight ${req.flightBooking} cancelled for trip ${req.tripId}`);
      },
    }
});

export type FlightsService = typeof flights;

File: go/patterns-use-cases/src/sagas/activities/flights.go
package activities

import (
	restate "github.com/restatedev/sdk-go"
	"log/slog"
)

type FlightBookingRequest struct {
	FlightId      string `json:"flight_id"`
	PassengerName string `json:"passenger_name"`
}

type Flights struct{}

func (Flights) Reserve(ctx restate.Context, _flightBookingRequest FlightBookingRequest) (string, error) {
	flightBookingId := restate.Rand(ctx).UUID().String()
	slog.Info("Flight reserved: " + flightBookingId)
	return flightBookingId, nil
}

func (Flights) Confirm(ctx restate.Context, flightBookingId string) error {
	slog.Info("Flight confirmed: " + flightBookingId)
	return nil
}

func (Flights) Cancel(ctx restate.Context, flightBookingId string) error {
	slog.Info("Flight cancelled: " + flightBookingId)
	return nil
}
